components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+                    }
+                },
+                _ => return PropertyRegistrationResult::SyntaxError,
+            }
+        },
+        None if matches!(syntax, Syntax::Anything) => None,
+        // initialValue is required if the syntax is not '*'.
+        _ => return PropertyRegistrationResult::SyntaxError,
+    };
+
+    let result =
+        registered_property_set
+        .register_property(Registration {
+            name: name.into(),
+            syntax: syntax,
+            inherits: inherits,
+            initial_value: initial_value,
+        });
+
+    match result {
+        Ok(_) => PropertyRegistrationResult::Ok,
+        Err(_) => PropertyRegistrationResult::InvalidModificationError,
+    }
+}
+
+/// Attempt to unregister a custom property.
+///
+/// This is used by the CSS.registerProperty implementations for Servo and Gecko
+/// in order to share logic. The caller should handle the returned
+/// PropertyRegistraitonResult by throwing the appropriate DOM error.
+pub fn unregister_property(
+    registered_property_set: &mut RegisteredPropertySet,
+    name: &str
+) -> PropertyRegistrationResult {
+    let name = match custom_properties::parse_name(name) {
+        Ok(name) => name,
+        Err(()) => return PropertyRegistrationResult::SyntaxError,
+    };
+
+    let result =
+        registered_property_set
+        .unregister_property(&name.into());
+
+    match result {
+        Ok(_) => PropertyRegistrationResult::Ok,
+        Err(_) => PropertyRegistrationResult::NotFoundError,
+    }
+}
+
+
+
+/// A CSS <custom-ident>.
+///
+/// We make a newtype for Atom and implement ToCss ourselves because the
+/// ToCss implementation for atom in `style_traits::values` uses `cssparsers`'s
+/// `serialize_string` function, which writes a double-quoted CSS string. We're
+/// only storing <custom-idents>, which should be serialized as specified.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct Ident(pub Atom);
+
+impl ComputedValueAsSpecified for Ident {}
+
+impl ToCss for Ident {
+    #[cfg(feature = "servo")]
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str(&*self.0)
+    }
+
+    #[cfg(feature = "gecko")]
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str(&self.0.to_string())
+    }
+}
+
+impl Deref for Ident {
+    type Target = Atom;
+
+    fn deref(&self) -> &Atom {
+        &self.0
+    }
+}
+
+/// A computed <url> value.
+///
+/// FIXME: While as_str() resolves URLs on the Servo side, it currently does not
+/// resolve them on the Gecko side (see the comment on Gecko's
+/// specified::url::SpecifiedUrl implementation). That means we don't actually
+/// resolve URLs when running in Stylo yet.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ComputedUrl(pub String);
```

Comment:
All of these variants duplicate names, to do so otherwise would make things very ugly (`properties_and_values::SpecifiedValue::PropertiesAndValuesSpecifiedAngle`?)